src/graphic/Fast3D/gfx_metal.cpp

```diff
@@ -0,0 +1,1041 @@
+    bool has_depth_buffer;
+    uint32_t msaa_level;
+    bool render_target;
+
+    // State
+    bool has_ended_encoding;
+    bool has_bounded_vertex_buffer;
+    bool has_bounded_fragment_buffer;
+
+    struct ShaderProgramMetal* last_shader_program;
+    MTL::Texture* last_bound_textures[2];
+    MTL::SamplerState* last_bound_samplers[2];
+
+    int8_t last_depth_test = -1;
+    int8_t last_depth_mask = -1;
+    int8_t last_zmode_decal = -1;
+};
+
+struct FrameUniforms {
+    simd::int1 frameCount;
+    simd::float1 noiseScale;
+};
+
+struct CoordUniforms {
+    simd::uint2 coords[MAX_PIXEL_DEPTH_COORDS];
+};
+
+static struct {
+    // Elements that only need to be setup once
+    SDL_Renderer* renderer;
+    CA::MetalLayer* layer; // CA::MetalLayer*
+    MTL::Device* device;
+    MTL::CommandQueue* command_queue;
+
+    int current_vertex_buffer_pool_index = 0;
+    MTL::Buffer* vertex_buffer_pool[kMaxVertexBufferPoolSize];
+    std::unordered_map<std::pair<uint64_t, uint32_t>, struct ShaderProgramMetal, hash_pair_shader_ids>
+        shader_program_pool;
+
+    std::vector<struct TextureDataMetal> textures;
+    std::vector<FramebufferMetal> framebuffers;
+    FrameUniforms frame_uniforms;
+    CoordUniforms coord_uniforms;
+    MTL::Buffer* frame_uniform_buffer;
+
+    uint32_t msaa_num_quality_levels[METAL_MAX_MULTISAMPLE_SAMPLE_COUNT];
+
+    // Depth querying
+    MTL::Buffer* coord_uniform_buffer;
+    MTL::Buffer* depth_value_output_buffer;
+    size_t coord_buffer_size;
+    MTL::Function* depth_compute_function;
+
+    // Current state
+    struct ShaderProgramMetal* shader_program;
+    CA::MetalDrawable* current_drawable;
+    std::set<int> drawn_framebuffers;
+    NS::AutoreleasePool* frame_autorelease_pool;
+
+    int current_tile;
+    uint32_t current_texture_ids[2];
+
+    uint32_t render_target_height;
+    int current_framebuffer;
+    size_t current_vertex_buffer_offset;
+    FilteringMode current_filter_mode = FILTER_THREE_POINT;
+
+    int8_t depth_test;
+    int8_t depth_mask;
+    int8_t zmode_decal;
+} mctx;
+
+// MARK: - Helpers
+
+static MTL::SamplerAddressMode gfx_cm_to_metal(uint32_t val) {
+    switch (val) {
+        case G_TX_NOMIRROR | G_TX_CLAMP:
+            return MTL::SamplerAddressModeClampToEdge;
+        case G_TX_MIRROR | G_TX_WRAP:
+            return MTL::SamplerAddressModeMirrorRepeat;
+        case G_TX_MIRROR | G_TX_CLAMP:
+            return MTL::SamplerAddressModeMirrorClampToEdge;
+        case G_TX_NOMIRROR | G_TX_WRAP:
+            return MTL::SamplerAddressModeRepeat;
+    }
+
+    return MTL::SamplerAddressModeClampToEdge;
+}
+
+// MARK: - ImGui & SDL Wrappers
+
+bool Metal_IsSupported() {
+    return MTLCopyAllDevices()->count() > 0;
+}
+
+bool Metal_Init(SDL_Renderer* renderer) {
+    mctx.renderer = renderer;
+    NS::AutoreleasePool* autorelease_pool = NS::AutoreleasePool::alloc()->init();
+
+    mctx.layer = (CA::MetalLayer *)SDL_RenderGetMetalLayer(renderer);
```

Comment:
```suggestion
    mctx.layer = (CA::MetalLayer *)SDL_RenderGetMetalLayer(renderer);
```
this should make format happy